### ISSUE TITLE
GPUVM fault in Mask-RCNN

### ISSUE THREAD

**Issue by repletetop** [ROCmSoftwarePlatform/tensorflow#31]
用WX9100,测试Mask-RCNN,
use  amd WX9100 graphics card 
tensorflow
test Mask-rcnn

name: Device 6861
AMDGPU ISA: gfx900
memoryClockRate (GHz) 1.5
pciBusID 0000:83:00.0
Total memory: 15.98GiB
Free memory: 15.73GiB

Ubuntu 16.04.1
rocm1.7
tensorflow-1.3.0-cp27-cp27mu-linux_x86_64.whl

运行到下面出错：
run this:
  detections, mrcnn_class, mrcnn_bbox, mrcnn_mask, \
            rois, rpn_class, rpn_bbox =\
            self.keras_model.predict([molded_images, image_metas], verbose=0)
错误提示：
error msg:
2018-04-03 15:39:00.781756: I tensorflow/core/kernels/conv_ops.cc:670] running auto-tune for Convolve
Memory access fault by GPU node-2 on address 0x32000. Reason: Page not present or supervisor privilege. Aborted (core dumped)
是什么原因？

**Comment by whchung** [comment]
從 log 上表示, 有一個 GPU kernel 存取了不存在的記憶體位置 (0x32000). 我需要重現本問題好進行後續調查.

請提供:
1. Mask-RCNN URL
1. 相依的 PIP 列表
1. 執行方式

**Comment by whchung** [comment]
@repletetop 我另外發現你的環境是 ROCm `1.7`, 而非 `1.7.1`. 可以請你提供下列資訊嗎:

```
apt list --installed | grep -E 'rock|compute-firmware'
```

其結果應該要是像:
```
compute-firmware/xenial,now 1.7.17 all [installed]
rock-dkms/now 1.7.137-ubuntu all [installed,upgradable to: 1.7.139-ubuntu]
```

**Comment by whchung** [comment]
@repletetop Also I'd like to request to switch to English so others on the forum can understand our conversation and might step in for help.

**Comment by briansp2020** [comment]
I'm having this issue as well. My configuration is default ROCm 1.7.1+tensorflow 1.3 image at GPUeater.com, updated with apt last night. (it seems there was new MIOpen-hip release), I can't provide more detail since my instance locked up and I can't log in any more...

**Comment by whchung** [comment]
@briansp2020 could you help provide some basic information so I can try reproduce the issue? thanks.

1. Mask-RCNN URL
1. dependent PIP packages
1. steps to reproduce the issue

**Comment by briansp2020** [comment]
Use GPUeater.com ROCm1.7.1+tensorflow1.3 image.
apt update & apt upgrade (this upgrades MIOpen-hip)
install some python dependencies (I forgot exactly what they were but running the notebook should tell you)
git clone https://github.com/matterport/Mask_RCNN
run train_shapes.ipynb.
it will hang.

I have modified the code some what to participate in https://www.kaggle.com/c/data-science-bowl-2018 and that code crashes giving the same error as OP. e-mail me (brianspark at hotmail.com) and I can send you the code.

**Comment by whchung** [comment]
@briansp2020 thanks. Let me try reproduce with original train_shapes.ipynb in `Mask_RCNN` first.

**Comment by whchung** [comment]
@repletetop @briansp2020 I'm able to reproduce the issue. trace points to a GPU kernel. I'll look into it.

**Comment by briansp2020** [comment]
@whchung  
Good. Hope its a quick fix. I'll be anxiously waiting.

**Comment by whchung** [comment]
@briansp2020 Luckily it seems to be a different issue than #16 / #21 / #23 / #34 where an exception within `MIOpen` somehow gets uncaught. The failing op is inside a custom CUDA / HIP kernel within TensorFlow and has nothing to do with `MIOpen`.

To help unblock you, I'm wondering are you able to build TensorFlow yourself? Forcing the op to be executed on CPU could be an workaround for now. I'll provide steps after I make sure it's possible.

**Comment by whchung** [comment]
failing GPU kernel collected with the following env vars:
```
HCC_SERIALIZE_KERNEL=3
HIP_TRACE_API=2
MIOPEN_ENABLE_LOGGING=1
HCC_DB=0x48A
```

```
^[[32m<<hip-api tid:12.3499 12.3499 hipLaunchKernel '_ZN10tensorflow12_GLOBAL__N_125ResizeNearestNeighborNHWCIfLb0EEEviPKT_iiiiiffPS2_' gridDim:{65536,1,1} groupDim:{1024,1,1} sharedMem:+0 stream:0.1 @85768606919508^[[0m
   hcc-cmd tid:12   pushing #0.1.7364 completion_signal=0  commandKind=hcCommandKernel <unknown_kernel>
   hcc-cmd tid:12   pushing #0.1.7365 completion_signal=0  commandKind=hcCommandMarker
   hcc-aql tid:12  barrier_aql #0.1.7365 header=1503(type=3,barrier=1,acquire=2,release=2) dep_signal[0]=0x0 dep_signal[1]=0x0 dep_signal[2]=0x0 dep_signal[3]=0x0 dep_signal[4]=0x0 completion_signal=0x553d000
   hcc-aql tid:12  dispatch_aql #0.1.7364(hwq=0x579d000) kernargs=72 header=b02(type=2,barrier=1,acquire=1,release=1) setup=3 grid=[65536.1.1] group=[1024.1.1] private_seg_size=0 group_seg_size=0 kernel_object=0x603f46200 kernarg_address=0x0x3b07200 completion_signal=0x5581000
   hcc-kernarg tid:12 kernarg_address: 0x0x3b07200, total of 72 bytes:
      0x00000000x3b07200: 0x00020000  0x00000000  0x2dedce00  0x00000006
      0x00000000x3b07210: 0x00000004  0x00000000  0x00000004  0x00000000
      0x00000000x3b07220: 0x00000100  0x00000000  0x00000008  0x00000000
      0x00000000x3b07230: 0x00000008  0x00000000  0x3f000000  0x3f000000
      0x00000000x3b07240: 0x2d801600  0x00000006
Memory access fault by GPU node-1 on address 0x3d000. Reason: Page not present or supervisor privilege.
```

**Comment by whchung** [comment]
+ @gstoner  for awareness as it yet another kernel driver level issue.

A workaround is available and it involve 2 steps:

1. `HSA_ENABLE_SDMA=0`
1. force `resize_nearnest_neighbor` op on CPU

The first one is a kernel driver issue, it's the same as #35 .

The second one seems to be an issue in AMDGPU compiler backend. I can work on the second one and will need @gstoner 's help on the first one.

**Comment by whchung** [comment]
+ @pfultz2 for awareness as it reveals at training phase we may yet bump into `MIOpen` exception unhandled issue depicted in #16 / #21 / #23 / #34.

In training phase `train_shapes.ipynb`, after apply workaround for inference (denoted above), we'll run into unhandled exception in `MIOpen`. I'll try provide more information.

**Comment by whchung** [comment]
@pfultz2 / @dagamayank here are the relevant info for somehow handled exception within `MIOpen`:

With the following workarounds:

1. `HSA_ENABLE_SDMA=0`
1. force `resize_nearest_neighbor` op on CPU

With the following env vars:

```
HSA_ENABLE_SDMA=0
HCC_SERIALIZE_KERNEL=3
HIP_TRACE_API=2
MIOPEN_ENABLE_LOGGING=1
LC_ALL=C
```

We 100% crashes `MIOpen` in this line in `train_shapes.ipynb`:

```
# Fine tune all layers
# Passing layers="all" trains all layers. You can also 
# pass a regular expression to select which layers to
# train by name pattern.
model.train(dataset_train, dataset_val, 
            learning_rate=config.LEARNING_RATE / 10,
            epochs=2, 
            layers="all")
```

Relevant logs:

```
MIOpen(HIP): miopenStatus_t miopenConvolutionBackwardDataGetWorkSpaceSize(miopenHandle_t, const miopenTensorDescriptor_t, const miopenTensorDescriptor_t, const miopenConvolutionDescriptor_t, const miopenTensorDescriptor_t, size_t *){
dyDesc = 0x8, 0x800, 0x4, 0x4
wDesc = 0x800, 0x400, 0x1, 0x1
convDesc = 0, 0, 0x2, 0x2, 0x1, 0x1,
dxDesc = 0x8, 0x400, 0x8, 0x8
workSpaceSize = 0x7fd1f54bdd02
}
MIOpen(HIP): miopenStatus_t miopenFindConvolutionBackwardDataAlgorithm(miopenHandle_t, const miopenTensorDescriptor_t, const void *, const miopenTensorDescriptor_t, const void *, const miopenConvolutionDescriptor_t, const miopenTensorDescriptor_t, const void *, const int, int *, miopenConvAlgoPerf_t *, void *, size_t, bool){
dyDesc = 0x8, 0x800, 0x4, 0x4
dy = 0x6605ef200
wDesc = 0x800, 0x400, 0x1, 0x1
w = 0x648e16c00
convDesc = 0, 0, 0x2, 0x2, 0x1, 0x1,
dxDesc = 0x8, 0x400, 0x8, 0x8
dx = 0x660a6f200
requestAlgoCount = 0x1
returnedAlgoCount = 0
perfResults =
workSpace = 0x66006f200
workSpaceSize = 0x180000
exhaustiveSearch = 0
}
terminate called after throwing an instance of 'miopen::Exception'
  what():  /home/whchung/MIOpen/src/mlo_dir_conv.cpp:105: No solution found
```

**Comment by whchung** [comment]
+ @bensander @scchan @pfultz2 @dagamayank for awareness.

The root cause has been figured out. Multiple instances of `libunwind` are loaded at runtime and messed up the exception handling routine. Need to disable `libunwind` dependency across various components within ROCm stack.

**Comment by whchung** [comment]
The issue would be resolved in the next ROCm 1.7.2 release where all `libunwind` dependencies be dropped. We will also publish a new set of PIP wheels.

**Comment by briansp2020** [comment]
Wow!
Thanks for the quick response. Looking forward to trying it out.

**Comment by whchung** [comment]
+ @acmeman925 for awareness. Turns out the issue in `resize_nearest_neighbor` TensorFlow op is on the host side logic, not in GPU kernels. Commit https://github.com/ROCmSoftwarePlatform/tensorflow/commit/efc1ef8beff229d92315c4908d64f5f9cc8803a9 fixed the issue wrt `resize_nearest_neighbor`.

Now to close the ticket properly, we still need:
- ROCm 1.7.2 release which strips `libunwind` from all major user-level components. The release shall be made available within this week.
- Updated PIP whl packages on `repo.radeon.com`. This would be done after ROCm 1.7.2 packages are online.
- `HSA_ENABLE_SDMA=0` be set on desktop machines. The issue won't be fixed in ROCm 1.7.2.

**Comment by whchung** [comment]
ROCm 1.7.2 has been released and PIP packages for ROCm 1.7.2 were also released.

Mask RCNN works fine with `HSA_ENABLE_SDMA=0`. Keep this ticket open for now.

**Comment by briansp2020** [comment]
I'm trying this with 1.7.2 and I'm getting different problem now. There has been some updates to Mask_RCNN code as well. So, I'm not sure whether this is new issues or not.

https://gist.github.com/briansp2020/97d4b1fc658c4340acfc2838208b1e2b

Used GPUeater and updated to 1.7.2 and install the whl from the repo today.

**Comment by briansp2020** [comment]
Actually, after upgrading numpy to the latest version, it seems to be working now.

**Comment by briansp2020** [comment]
Is there a way to check the load on the GPU? Unlike nvidia-smi, rocm-smi does not show load or memory usage...

**Comment by briansp2020** [comment]
After running training process for a while, I'm unable to shutdown the machine properly. I had this issue when I tried this the last time.
Output from my attempt to run my 2018 Data Science Bowl training code.
https://gist.github.com/briansp2020/dd49704da9d43eac3b429ed4a804c8c9

The training process takes a lot of resources and GPUeater instance does not have enough CPU power nor enough main memory. So, I suspect that is why I got error. But after getting the error, I could not shutdown the instance properly. I had to terminate the instance so I no longer have the drive image to reproduce the issue. :(

I wish I had a Vega FE so I can experiment locally. ;) @gstoner

**Comment by briansp2020** [comment]
ROCm 1.8 is out. Is HSA_ENABLE_SDMA=0 still required? What does SDMA do? Would it increase performance if working properly?

**Comment by whchung** [comment]
Unfortunately SDMA issue won’t be resolved until ROCm 1.9.

**Comment by briansp2020** [comment]
Tried the latest shapes demo using GPUeater Vega FE ROCm1.8+tensowflow1.3 image and I'm getting errors.
https://gist.github.com/briansp2020/4d2a0b53bdd422ab2259016dd4f438ce

@whchung did it work for you? There has been some updates to Mask_RCNN code and I'm using the latest. But I don't think there are changes that would cause issues. The same code runs on my home machine using tensorflow 1.8 on GTX1080.

**Comment by whchung** [comment]
/cc @parallelo for awareness

@briansp2020 we can run networks such as Mask RCNN, YOLO v3, PSPNET properly on TensorFlow 1.3 + ROCm `1.7.2`, under the premise that `HSA_ENABLE_SDMA=0` is set. They work fine either executed directly from python, or within jupyter notebook.

That said, I haven't personally tested ROCm `1.8`. Test coverage for ROCm releases doesn't cover Mask RCNN yet.

**Comment by briansp2020** [comment]
It worked on ROCm 1.7.2 + TF 1.3 as I reported previously. I think ROCm 1.8 + TF 1.3 broke it. Then again, for ROCm 1.8, I used GPUeater default image where as for 1.7.2, I updated their 1.7.1 set up manually myself.

I did set HSA_ENABLE_SDMA=0. If someone from ROCm team can double check, I'd appreciate it.
Thanks!

**Comment by parallelo** [comment]
I don't recall seeing any new memory access fault-related regressions in 1.8.  But like Jack mentioned, we haven't started testing Mask RCNN regularly yet.  

Regarding SDMA, I believe there will be some upcoming ROCr fixes related to the atomic path for SDMA.  The associated problem was an infrequent random hang with hipCaffe; so, the symptom seems different than reported in this ticket.

**Comment by whchung** [comment]
@parallelo could weigh in, but we are getting MLperf ( https://mlperf.org/ ) integrated in the CI process and Mask RCNN is part of the benchmark.

**Comment by briansp2020** [comment]
Object detection in MLperf (https://github.com/mlperf/reference) is using Mask RCNN written in Caffe2.

**Comment by briansp2020** [comment]
It looks like GPUeater has old tensorflow. I uninstalled and install from http://repo.radeon.com/rocm/misc/tensorflow/tensorflow-1.3.0-cp35-cp35m-linux_x86_64.whl

**Comment by repletetop** [comment]
Thanks @whchung , 我更换了N卡
